search/views.py

```diff
@@ -318,12 +121,109 @@ def search(request):
         logger.error('Could probably not connect to Solr - %s' % e)
         tvars.update({'error_text': 'The search server could not be reached, please try again later.'})
 
+    # enables AJAX clustering call & html clustering facets rendering
+    if settings.ENABLE_SEARCH_RESULTS_CLUSTERING:
+        tvars.update({'clustering_on': True})
+
     if request.GET.get("ajax", "") != "1":
         return render(request, 'search/search.html', tvars)
     else:
         return render(request, 'search/search_ajax.html', tvars)
 
 
+def _get_ids_in_cluster(request, requested_cluster_id):
+    """Get the sound ids in the requested cluster. Used for applying a filter by id when using a cluster facet.
+    """
+    try:
+        requested_cluster_id = int(requested_cluster_id) - 1
+    
+        # results are cached in clustering_utilities, available features are defined in the clustering settings file.
+        result = cluster_sound_results(request, features=DEFAULT_FEATURES)
+        results = result['result']
+
+        sounds_from_requested_cluster = results[int(requested_cluster_id)]
+
+    except ValueError:
+        return []
+    except IndexError:
+        return []
+
+    return sounds_from_requested_cluster
+
+
+def clustering_facet(request):
+    """Triggers the computation of the clustering, returns the state of processing or the clustering facet.
+    """
+    # pass the url query params for later sending it to the clustering engine
+    url_query_params_string = request.META['QUERY_STRING']
+    # remove existing cluster facet filter from the params since the returned cluster facets will include 
+    # their correspondinng cluster_id query parameter (done in the template)
+    url_query_params_string = re.sub(r"(&cluster_id=[0-9]*)", "", url_query_params_string)
+
+    result = cluster_sound_results(request, features=DEFAULT_FEATURES)
+
+    # check if computation is finished. If not, send computation state.
+    if result['finished']:
+        if result['result'] is not None:
+            results = result['result']
+            num_clusters = num_clusters = len(results) + 1
+        else:
+             return JsonResponse({'status': 'failed'}, safe=False)
+    elif result['error']:
+        return JsonResponse({'status': 'failed'}, safe=False)
+    else:
+        return JsonResponse({'status': 'pending'}, safe=False)
+
+    num_sounds_per_cluster = [len(cluster) for cluster in results]
+    classes = {sound_id: cluster_id for cluster_id, cluster in enumerate(results) for sound_id in cluster}
+
+    # label clusters using most occuring tags
+    sound_instances = sounds.models.Sound.objects.bulk_query_id(map(int, classes.keys()))
+    sound_tags = {sound.id: sound.get_sound_tags() for sound in sound_instances}
```

Comment:
Oh no! Don't call `get_sound_tags` as this will make an extra query for each sound. Because you used `bulk_query_id`, the sound objects are already loaded with an added `tag_array` which contains a list of the tags. Therefore you can do `sound_tags = {sound.id: sound.tag_array for sound in sound_instances}` and will save many queries.

This makes me think however that the `get_sound_tags` method could be updated and first check if `tag_array` exists in the model. Could be refactored to something like:

```python

    def get_sound_tags(self, limit=None):
        """
        Returns the tags assigned to the sound as a list of strings, e.g. ["tag1", "tag2", "tag3"]

        :param limit: The maximum number of tags to return
        """
        if hasattr(self, 'tag_array'):
            return self.tag_array[0:limit]
        else:
            return [ti.tag.name for ti in self.tags.select_related("tag").all()[0:limit]]
```

(docstring should be updated as well)

I'd suggest that you make this improvement to `get_sound_tags` and then you don't need to change the code here. Adding a unit test to check that using `get_sound_tags` triggers no queries if sound is loaded using `bulk_query_id` method is also a good idea :)

Reply:
Wow, this `get_sound_tags` was actually making these extra queries and making the cluster facets take a lot of time to appear on the page. Thanks!